qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -192,26 +228,24 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
             (for instance, when the specified synthesis method is not available).
         """
 
-        # In the future this pass will be recursive as we may have annotated gates
-        # whose definitions consist of other annotated gates, whose definitions include
-        # LinearFunctions. Note that in order to synthesize a controlled linear
-        # function, we must first fully synthesize the linear function, and then
-        # synthesize the circuit obtained by adding control logic.
-        # Additionally, see https://github.com/Qiskit/qiskit/pull/9846#pullrequestreview-1626991425.
-
-        # If there are no high level operations / annotated gates to synthesize, return fast
-        hls_names = set(self.hls_plugin_manager.plugins_by_op)
-        node_names = dag.count_ops()
-        if "annotated" not in node_names and not hls_names.intersection(node_names):
-            return dag
```

Comment:
I realize there isn't a quick lookup anymore because the equivalence libraries' lookup keys are not pure strings anymore. But it's a real shame to lose this, because it makes the pass much faster in the case when everything is in standard library gates (especially for really large circuits). I think if we did:

```
gates = {node[0] for node in self._equiv_lib.graph.nodes()}
```
or 
```
gates = {node[0] for node in self._equiv_lib._key_to_node_index}
```
That will give us a list of gate names in the equivalence library and we could retain this (along with a corresponding target and/or basis gates lookup). What do you think?

Reply:
An update after thinking about this: intuitively, to restore the previous optimization we need to compute
1. A = list of names `dag.count_ops()`
2. B = list of names in equivalence library (the second suggested command works)
3. C = list of names supported by the target. Looking at the code, I believe it should be something like `self._target._gate_map.keys()`
4. D = list of names in `self.hls_plugin_manager.plugins_by_op`

and check if `A` is a subset of `(B | C) \ D`.

However I believe we have the problem for open-controlled gates (as per the code adapted from `UnrollCustomDefinitions`): a gate's name might be supported by the target or belong to equivalence library, yet because it's open-control we actually can't skip synthesis and use the `definition`. Any suggestions? 

If this is not a trivial task, maybe we can postpone restoring this optimization to a follow-up PR?

Reply:
Yeah, I think it's fine to defer it, I didn't have a good answer for this either in `UnrollCustomDefinitions` for #10788 we could potentially implement this a performance fix for 0.45.0 post rc1 if needed.